rs-matter/src/data_model/core.rs

```diff
@@ -15,142 +15,759 @@
+                let more_chunks = req
+                    .respond(&self.handler, Some(id), &mut attrs, &mut wb, false)
+                    .await?;
+
+                exchange.send(OpCode::ReportData, wb.as_slice()).await?;
+
+                if !Self::recv_status(exchange).await? {
+                    info!("Subscription {node_id:x}::{id} removed during reporting");
+                    return Ok(false);
+                }
+
+                if !more_chunks {
+                    break;
                 }
             }
         }
 
+        Ok(true)
+    }
+
+    async fn rx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        let buffer = self.tx_buffer(exchange).await?;
+
+        if let Some(mut buffer) = buffer {
+            let rx = exchange.rx()?;
+
+            buffer.clear();
+            buffer.extend_from_slice(rx.payload()).unwrap();
+
+            exchange.rx_done()?;
+
+            Ok(Some(buffer))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn tx_buffer(&self, exchange: &mut Exchange<'_>) -> Result<Option<B::Buffer<'a>>, Error> {
+        if let Some(mut buffer) = self.buffers.get().await {
+            buffer.resize_default(MAX_EXCHANGE_TX_BUF_SIZE).unwrap();
+
+            Ok(Some(buffer))
+        } else {
+            Self::send_status(exchange, IMStatusCode::Busy).await?;
+
+            Ok(None)
+        }
+    }
+
+    async fn recv_status(exchange: &mut Exchange<'_>) -> Result<bool, Error> {
+        let rx = exchange.recv().await?;
+        let opcode = rx.meta().proto_opcode;
+
+        if opcode == OpCode::StatusResponse as u8 {
+            let resp = StatusResp::from_tlv(&get_root_node_struct(rx.payload())?)?;
+
+            if resp.status == IMStatusCode::Success {
+                Ok(true)
+            } else {
+                warn!(
+                    "Got status response {:?}, aborting interaction",
+                    resp.status
+                );
+
+                drop(rx);
+                exchange.acknowledge().await?;
+
+                Ok(false)
+            }
+        } else {
+            Err(ErrorCode::Invalid.into())
+        }
+    }
+
+    async fn send_status(exchange: &mut Exchange<'_>, status: IMStatusCode) -> Result<(), Error> {
+        exchange
+            .send_with(|_, wb| {
+                StatusResp::write(wb, status)?;
+
+                Ok(Some(OpCode::StatusResponse.into()))
+            })
+            .await
+    }
+}
+
+impl<'a, const N: usize, B, T> ExchangeHandler for DataModel<'a, N, B, T>
+where
+    T: DataModelHandler,
+    B: BufferAccess<IMBuffer>,
+{
+    async fn handle(&self, exchange: &mut Exchange<'_>) -> Result<(), Error> {
+        DataModel::handle(self, exchange).await
+    }
+}
+
+impl<'a> ReportDataReq<'a> {
+    // This is the amount of space we reserve for other things to be attached towards
+    // the end of long reads.
+    const LONG_READS_TLV_RESERVE_SIZE: usize = 24;
```

Comment:
should we add unit of measurement (i.e. `_BYTES` suffix) to constants?

Reply:
Let's not do it. The constant is pre-existing btw. :)